android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/Utils.kt

```diff
@@ -11,80 +11,114 @@ private const val SEARCH_TYPE_SAP_FOLLOW_ON = "sap-follow-on"
-    var code: String? = null
+    var code: String? = "none"
 
-    if (provider.codeParam.isNotEmpty()) {
-        code = uri.getQueryParameter(provider.codeParam)
+    if (provider.codeParamName.isNotEmpty()) {
+        code = uri.getQueryParameter(provider.codeParamName)
         if (code.isNullOrEmpty() &&
-            provider.name == "baidu" &&
+            provider.telemetryId == "baidu" &&
             uri.toString().contains("from=")
         ) {
             code = uri.toString().substringAfter("from=", "")
                 .substringBefore("/", "")
         }
 
-        // For Bug 1751920
-        if (!validCodeSet.contains(code)) {
-            code = "other"
-        }
-
-        // Glean doesn't allow code starting with a figure
-        if (code != null && code.isNotEmpty()) {
-            val codeStart = code.first()
-            if (codeStart.isDigit()) {
-                code = "_$code"
-            }
-        }
+        val result = determineSearchTypeAndCode(code, provider, paramSet)
+        type = result.searchType
 
         // Try cookies first because Bing has followOnCookies and valid code, but no
         // followOnParams => would tracks organic instead of sap-follow-on
-        if (provider.followOnCookies.isNotEmpty()) {
+        if (!provider.followOnCookies.isNullOrEmpty()) {
             // Checks if engine contains a valid follow-on cookie, otherwise return default
             getTrackKeyFromCookies(provider, uri, cookies)?.let {
                 return it.createTrackKey()
             }
         }
 
         // For Bing if it didn't have a valid cookie and for all the other search engines
-        if (hasValidCode(uri.getQueryParameter(provider.codeParam), provider)) {
+        if (hasValidCode(uri.getQueryParameter(provider.codeParamName), provider)) {
             var channel = uri.getQueryParameter(CHANNEL_KEY)
 
             // For Bug 1751955
             if (!validChannelSet.contains(channel)) {
                 channel = null
             }
 
-            val type = getSapType(provider.followOnParams, paramSet)
-            return TrackKeyInfo(provider.name, type, code, channel).createTrackKey()
+            type = if (provider.followOnParamNames == null) {
+                SEARCH_TYPE_SAP
+            } else {
+                getSapType(provider.followOnParamNames, paramSet)
+            }
+            return TrackKeyInfo(provider.telemetryId, type, result.modifiedCode, channel).createTrackKey()
+        }
+    }
+
+    return TrackKeyInfo(provider.telemetryId, type, code).createTrackKey()
+}
+
+/**
+ * Determines type and code needed for key information for a SERP.
+ */
+data class SearchTypeAndCode(val searchType: String, val modifiedCode: String?)
+
+private fun determineSearchTypeAndCode(
+    code: String?,
+    provider: SearchProviderModel,
+    paramSet: MutableSet<String>,
+): SearchTypeAndCode {
+    var searchType = SEARCH_TYPE_ORGANIC
+    var modifiedCode = code
```

Comment:
IMO using mutable variables like this can make the function hard to reason about. I would prefer to set both of these as immutable declarations using `when` expressions. Since they're both reliant on the same logic, you could use Kotlin's `Pair` destructing and the `to` infix function to write something like this:

```
    if (code.isNullOrEmpty()) {
        return SearchTypeAndCode(SEARCH_TYPE_ORGANIC, code)
    }

    val (searchType, modifiedCode) = when {
        provider.taggedCodes.contains(code) -> {
            val searchType = if (!provider.followOnParamNames.isNullOrEmpty() &&
                provider.followOnParamNames.any { prefix ->
                    paramSet.contains(prefix)
                }
            ) {
                SEARCH_TYPE_SAP_FOLLOW_ON
            } else {
                SEARCH_TYPE_SAP
            }
            searchType to code
        }
        provider.organicCodes?.contains(code) == true -> SEARCH_TYPE_ORGANIC to code
        provider.expectedOrganicCodes?.contains(code) == true -> {
            "" to "none"
        }
        else -> {
            "" to "other"
        }
    }

    // Glean doesn't allow code starting with a figure
    val finalCode = modifiedCode.takeIf { !it.first().isDigit() } ?: "_$modifiedCode"
    return SearchTypeAndCode(searchType, finalCode)
```

I think this helps remove some of the mental overhead of tracking down where each variable gets initialized, since it guarantees that both are being set in each conditional case.